File: lowest common ancestor with parent pointer.py
class TreeNode(object):
  def __init__(self, val):
    self.val = val
    self.left = None
    self.right = None
    self.parent = None

class Solution(object):
  def lca(self, p, q):
    if not p or not q:
      return

    h1 = self.getHeight(p)
    h2 = self.getHeight(q)
    if h1 > h2: #1比2短
      h1, h2 = h2, h1
      p, q = q, p

    while h1 < h2:
      if not q:
        return
      q = q.parent
      h2 -= 1
    和linkedlist一样，先把两个node拉到同一个高度，再一起往前走
    while p and q:
      if p == q:
        return p
      p = p.parent
      q = q.parent
    return


  def getHeight(self, node):
    height = 0

    while node:
      node = node.parent
      height+=1
    return height

File: lowest common ancestor with parent pointer.cpp
//find lowest common ancestor with parent pointer. #fb #lca #tree #binarytree O(1) space and O(n) time.

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode *parent;
    TreeNode(int x) : val(x), left(NULL), right(NULL), parent(NULL) {}
};

int getHeight(TreeNode *node) {
    int height = 0;
    while (node) {
        node = node->parent;
        height++;
    }
    return height;
}

TreeNode *lca(TreeNode *node1, TreeNode *node2) {
    if (!node1 || !node2)
        return NULL;

    int height1 = getHeight(node1);
    int height2 = getHeight(node2);

    if (height1 > height2) {
        swap(height1, height2);
        swap(node1, node2);
    }

    while (height1 < height2) {
        if (!node2)
            return NULL;
        node2 = node2->parent;
        height2--;
    }

    while (node1 && node2) {
        if (node1 == node2)
            return node1;
        node1 = node1->parent;
        node2 = node2->parent;
    }

    return NULL;
}